mlflow/store/artifact/databricks_artifact_repo.py

```diff
@@ -714,11 +723,26 @@ def get_creds_and_upload(staged_upload_chunk):
             for src_file_path, upload_future in inflight_uploads.items():
                 try:
                     upload_future.result()
+                    if pbar:
+                        pbar.update()
+                        pbar.refresh()
                 except Exception as e:
                     failed_uploads[src_file_path] = repr(e)
 
+        if MLFLOW_ENABLE_ARTIFACTS_PROGRESS_BAR:
+            try:
+                from tqdm.notebook import tqdm
+
+                pbar = tqdm(total=len(staged_uploads), desc="Uploading artifacts")
+            except ImportError:
+                pbar = None
+        else:
+            pbar = None
```

Comment:
Can we use `ArtifactProgressBar` here?

Reply:
If we replace with ArtifactProgressBar, we can only iterate over the outer for loop `chunk_list(staged_uploads, _ARTIFACT_UPLOAD_BATCH_SIZE)` and update step as chunk_size, instead of updating in the inner loop of `for src_file_path, upload_future in inflight_uploads.items():` by step 1.

Reply:
Can we flatten the loop? For example:



```python
for a in range(10):
    for b in range(10):
        yield a, b


def flattened():
    for a in range(10):
        for b in range(10):
            yield a, b


for a, b in flattened():
    print(a, b)

```

Reply:
Just come up with a problem of `ArtifactProgressBar`, it just yields the item, even if the item's operation fails, say `future.result()` fails, it updates the progress bar anyways...

Reply:
> Can we flatten the loop? For example:
> 
> ```python
> for a in range(10):
>     for b in range(10):
>         yield a, b
> 
> 
> def flattened():
>     for a in range(10):
>         for b in range(10):
>             yield a, b
> 
> 
> for a, b in flattened():
>     print(a, b)
> ```

The outer loop is here https://github.com/mlflow/mlflow/blob/e53b94c2a29da01a81e5a8a65d657a6a3ccd220e/mlflow/store/artifact/databricks_artifact_repo.py#L720. But the inner loop is defined within this function `get_creds_and_upload` https://github.com/mlflow/mlflow/blob/e53b94c2a29da01a81e5a8a65d657a6a3ccd220e/mlflow/store/artifact/databricks_artifact_repo.py#L714. 

In your example, we have to yield `a` first, then we can get `b` in the inner loop.

Reply:
Can we do something like this?

```python
        def iter():
            for chunk in chunk_list(staged_uploads, _ARTIFACT_UPLOAD_BATCH_SIZE):
                write_credential_infos = self._get_write_credential_infos(
                    run_id=self.run_id,
                    paths=[staged_upload.dst_run_relative_artifact_path for staged_upload in chunk],
                )

                inflight_uploads = {}
                for staged_upload, write_credential_info in zip(chunk, write_credential_infos):
                    upload_future = self.thread_pool.submit(
                        self._upload_to_cloud,
                        cloud_credential_info=write_credential_info,
                        src_file_path=staged_upload.src_file_path,
                        dst_run_relative_artifact_path=staged_upload.dst_run_relative_artifact_path,
                    )
                    inflight_uploads[staged_upload.src_file_path] = upload_future

                for src_file_path, upload_future in inflight_uploads.items():
                    yield upload_future, src_file_path

        # Join futures to ensure that all artifacts have been uploaded prior to returning
        failed_uploads = {}
        for f, p in ArtifactsProgressBar.files(iter()):
            try:
                f.result()
            except Exception as e:
                failed_uploads[p] = repr(e)
```

Reply:
Is there a way we can catch the error happens for f.result() back to ArtifactProgressBar? And if there's an error we don't update the progress bar

Reply:
To do that, we need to await a future in `ArtifactProgressBar`